substrate/client/authority-discovery/src/worker.rs

```diff
@@ -224,6 +226,29 @@ where
 			None => None,
 		};
 
+		// Make sure we have only valid peer ids in public addresses.
+		let public_addresses_only = {
+			let local_peer_id: Multihash = network.local_peer_id().into();
+
+			config.public_addresses_only.map(|addresses| {
+				addresses
+					.into_iter()
+					.map(|mut address| {
+						if let Some(multiaddr::Protocol::P2p(peer_id)) = address.iter().last() {
+							if peer_id != local_peer_id {
+								error!(
+									target: LOG_TARGET,
+									"Discarding invalid local peer ID in public address {address}.",
+								);
+								address.pop();
```

Comment:
I wonder what would be the best way to handle this. 

Since these are passed by the node-operator as a CLI param, we could probably return an error high level (where we receive the params)?

Node operators could abouse this CLI param to publish other addreses in DHT, would it be safer to return an error high level and here to `filter_map` and ignore the addresses with different peer IDs (in case we missed anything along the way)?

Reply:
Probable makes sense returning an error high-level and terminating the node. I'll look if it's possible to implement it in a clean fashion.

As for abusing, don't think there is an abuse vector: even if `/p2p/...` protocol is not filtered (is it allowed to be passed as not the last protocol of the address?), the authority discovery records are signed with the publishing node's network key and the signatures a checked during `get_record` DHT call to match a peer id in the record.

What do you think?

Reply:
> Probably makes sense returning an error high-level and terminating the node. I'll look if it's possible to implement it in a clean fashion.

It doesn't play well with the current design of the nodes startup. It would require introducing a new error enum for node services in polkadot, cumulus, substrate and pulling a libp2p dependency for `Multiaddr` introspection if we want to filter it in a higher level code. We could also make the authority-discovery constructor fallible and return `Result` from there, but this would require new error types anyway.

So, generally, I don't think we should implement this as an invalid peer id in the public address poses no real threat, and can just be discarded with an error in the log.

Reply:
That makes sense, thanks for looking into this 🙏